ceph-volume/util: Improve performance for Ceph OSDs that use encryption

hydro-b opened the pull request:
The commits in this PR are the following:

Changes in ceph-volume/util/encryption.py and disk.py to influence parameters used when opening encrypted devices.

- minor formatting fix
~~- automatically derive the optimal sector size to use for cryptsetup LuksFormat~~
- automatically disable dm-crypt "workqueues" for flash media when opening encrypted devices to improve I/O performance

Author has proof of performance improvements on request (not sure if / how to supply in this PR). These changes have been discussed on [ceph-user ML](https://lists.ceph.io/hyperkitty/list/ceph-users@ceph.io/thread/D5URLVPVGX52O6WYV474T6IVTHGHCASB/) and pitched at ceph-performance weekly. @markhpc asked to write a PR, so here we are ;-).


## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

hydro-b commented:
jenkins test make check

baergj commented:
@hydro-b Unfortunately, that ML link you posted in the description is dead. Any chance you could post the results to the ticket?

baergj commented:
FWIW, I've run the workqueue-disabled changes (haven't tried the sector configuration) against a fairly performant nvme system in RBD configuration. My observations in this environment:
* At low load, latency does not appear affected much by these changes. Any differences don't really exceed our error bars.
* At high load that includes writes, the larger the write, the more effective this change is. For example, OSD node CPU usage dropped by almost half for 4m sequential writes, with substantial improvements in latency (p90 reads were two orders of magnitude faster).
* Large block read-only loads might have had slightly worse latency, but hard to tell for certain.

baergj commented:
Looks like cryptsetup 2.4.0 will automatically detect the best LUKS sector size: https://mirrors.edge.kernel.org/pub/linux/utils/cryptsetup/v2.4/v2.4.0-ReleaseNotes

baergj commented on src/ceph-volume/ceph_volume/util/encryption.py:
```diff
@@ -60,6 +61,63 @@ def get_sector_size(device):
+
+    The encrypted device will come online without the options active.
+    """
+    noworkqueue_support = False
```
This variable appears to be set only and never read.

hydro-b replied:
This variable was removed in updated PR.

baergj replied:
👍

hydro-b commented:
@baergj Thanks for looking into this. I will post test results in this issue. I have done a ton of extra testing in the mean time (also with different sector sizes). There is at least a flaw in the logic regarding the sector size to use when formatting. I will update the PR accordingly and check your other comments.

baergj commented:
@hydro-b You're welcome, and thanks! Note that I'm just an interested community member, so don't take any of my comments for more than they're worth. :)

baergj commented on src/ceph-volume/ceph_volume/util/encryption.py:
```diff
@@ -60,6 +61,63 @@ def get_sector_size(device):
+    if "--perf-no_read_workqueue" in cryptsetup_help:
+        noworkqueue_support=True
+
+        if is_rotational(device) is False:
```
To be more Pythonic, this should be `if not is_rotational(device):`. But given the above comment about the unused var, this could probably all be simplified to something like:
```python
     if "--perf-no_read_workqueue" not in cryptsetup_help or is_rotational(device):
         return []

     return ['--perf-no_read_workqueue', '--perf-no_write_workqueue']
```

hydro-b replied:
Yes, this has been rewritten as you suggested.

baergj replied:
Thanks!

baergj commented:
Given that Reef will likely be on CentOS 9, which comes with cryptsetup 2.4.3 as far as I can tell, my recommendation would be to drop the sector size logic from this PR and just focus on disabling the workqueues, unless you think that the cryptsetup 2.4+ sector size logic can be improved on? (In which case separating the PRs might make sense, since the benefit from the workqueue logic is more clear.)

hydro-b commented:
@baergj Sorry for responding so late. I didn't have time to look into this past couple of weeks. Until yesterday. I've applied your suggestions, and moved the "is_rotational" function to "disk.py", which seems a more appropriate place (similar functions in that file). 

Regarding cryptsetup automatically detecting proper sector size: I have tested this, and it works indeed like expected. So I will remove that code from this PR. I have made a backport for Ubuntu 20.04 (focal) based on 22.04 cryptsetup (2.4.3). That works.

Since yesterday we are running this in production on three nodes, 4K sector size, with both workqueues bypassed on Ubuntu Focal 20.04 with 5.15 kernel. All seems to be running smooth.

Still need time to find time to condense all our benchmarks into something comprehensible. And figure out now I can rebase my branch based on master and fixup the PR. I'll try to make progress the coming days or directly after Cephalacon.

anthonyeleven commented:
Note:  this may resolve https://tracker.ceph.com/issues/58569

anthonyeleven commented on src/ceph-volume/ceph_volume/util/disk.py:
```diff
@@ -911,6 +931,10 @@ def get_devices(_sys_block_path='/sys/block', device=''):
+        fallback_rotational = '1'
+        metadata['rotational'] = get_file_contents(sysdir +
+                                                   "/queue/rotational",
+                                                   fallback_rotational)
```
How does this relate to the above `is_rotational`?

hydro-b replied:
There are already some fallbacks in place (i.e. sector size) so I added this one as well. This *should* not be needed, but does not hurt.

baergj replied:
I think Anthony's question was more about the fact that `is_rotational()` and `metadata['rotational']` don't seem to be related to each other. Maybe they should be?

anthonyeleven replied:
Indeed, I discovered the multiple implementations with our RGW Lua work a while back.  I can see where multiple ways of determining HDD vs SSD might return different results, leading to even more confusion.

anthonyeleven reviewed (commented):
I'm not seeing where this touches sector/block size of devices.  Is that implicit somehow?  Also, does that only address devices with 4KB blocks, or will it adapt equally to those with 8KB, 16KB, even 64KB block/IU sizes?  Reference https://github.com/ceph/ceph/pull/43691 which enables aligning `min_alloc_size` to a kernel-provided optimal value.

baergj commented:
@anthonyeleven I think the sector size changes might have been dropped based on:
* https://github.com/ceph/ceph/pull/49554#issuecomment-1462788225
* https://github.com/ceph/ceph/pull/49554#issuecomment-1472231395

anthonyeleven commented:
Ack, thanks.  Perhaps the PR title should be amended to reflect that.

hydro-b commented:
> Ack, thanks. Perhaps the PR title should be amended to reflect that.

Yup, title has been fixed.

hydro-b commented on src/ceph-volume/ceph_volume/util/disk.py:
```diff
@@ -195,6 +195,26 @@ def device_family(device):
+                rotational = False
+
+    return rotational
+
```
> `src/common/blkdev.cc` and `src/ceph-volume/ceph_volume/util/device.py` and `src/pybind/mgr/rook/rook_cluster.py` have existing functions that serve this purpose. Is this wheel reinvention?

Note that  `src/ceph-volume/ceph_volume/util/device.py`  includes `disk.py`. AFAICT it's not duplicating any effort here.

I have taken a look at `src/pybind/mgr/rook/rook_cluster.py` , but it does not look like it's performing similar functions as `disk.py`.

I agree that a common library, shared between rook/cephadm/ceph-volume, might make sense. But that seems to be something that has to be discussed between maintainers of those projects

> Are we doing it the same way across the code?

Yes, this is how it's done for other disk properties as well.
 
> Might inspecting `/sys/block` be cleaner than shelling out to `lsblk`? Can we count on `lsblk` being always installed and in the path?

If `lsblk` is doing a good job Ceph does not have to reinvent the wheel here. It's used in ceph-volume in lots of places. I have not checked what distro's or containers come with lsblk by default. It should be a dependency of ceph-osd. I checked on ubuntu, and it's not listed as a dependency there. Containers (created by Ceph) will have this command available.  I do want to stress however, that the use of `lsblk` here does not add any new dependencies, it's just another part of the code that makes use of it. And that the general question if that is a good thing or not should be discussed on a higher level than this PR.

anthonyeleven replied:
I wonder how each implementation behaves on top of LVM, dmcrypt, and when containerized.  I would not be surprised if there are clusters out there crippled by specious determinations of rotation.

hydro-b replied:
> This is a good catch, Anthony.
> 
> @hydro-b It looks like there are now three ways in ceph-volume that the rotational state of a drive is inspected; the first two pre-existed your changes:
> 
>     1. A property returned from disk.py's `get_devices()`
> 
>     2. device.py's `Device.rotational` (this seems like the most complete implementation)
> 
>     3. The new `is_rotational()` added here
> 
> 
> Do you think there's a way to unify the implementations of 2 and 3, or, per the other comment, at least 1 and 3? At a glance I'm not sure how this would be done without a few rounds of refactoring.

I've looked at using 1) and 2).

1) disk.py's `get_devices()` excludes logical volumes
2) device.py's rejects the LVM volume it gets fed. It ends with a general rejected reason. IIRC the issue has to do with /dev/vg/lv being a symlink to a device mapper device, the format of the symlink: `/dev/vg/lv` instead of `vg.lv`. I've hacked up a `lsblk` version does a regex match and fixes the device into what lsblk wants. And then my time to work on this was up and had to shift to operational stuff.

So yeah, I agree that a single function to return rotational state for whatever device type / format gets passed would be nice, however it's not there yet.

hydro-b commented:
> @hydro-b Could you take a look at comments I added on Mar 9 and 10? I think once those are addressed this is probably ready to merge

Yes, I believe I have addressed all comments now. I noticed that in the rebased version if pushed earlier part of the code was duplicated, and that one remnant was still there. That's me using meld in a 3-way conflict resolution workflow for the first time (and obviously did not see this in time). I have rebased and forced pushed again. Hopefully all looks good now.

hydro-b commented:
@guits AFAIK you are the maintainer of ceph-volume. Would you mind reviewing this PR?

hydro-b commented:
@markhpc: Would you be able to do performance tests with these read/write workqueues bypassed (--perf-no_read_workqueue --perf-no_write_workqueue as arguments for cryptsetup)?

markhpc commented:
@hydro-b I'm afraid my schedule is pretty booked these days.  @vumrao @pcuzner Is this something the DFG can help out with?

anthonyeleven commented:
bump

markhpc commented:
@hydro-b Heya, are you still updating this PR?  Just wanted to see if you saw @anthonyeleven's feedback.  We've got a couple of different users that would benefit from this work.  Thanks!

hydro-b commented:
> @hydro-b Heya, are you still updating this PR? Just wanted to see if you saw @anthonyeleven's feedback. We've got a couple of different users that would benefit from this work. Thanks!

See my reply on @anthonyeleven and @baergj comments. I also would like to see this support being merged. Maybe @guits has a viewpoint about this? I'm not sure what the best fix / approach would be. The more changes are made to general functions, the greater the chance of regressions.

anthonyeleven commented:
That's why I suggested the route of allowing arguments to be passed through vs hardcoding them.

hydro-b commented:
> That's why I suggested the route of allowing arguments to be passed through vs hardcoding them.

Not sure if I follow you. Can you elaborate? Would you prefer a "config option" instead of having the code figure out what would be the best thing to do?

anthonyeleven commented:
Yes, I think that would be flexible, at least wrt the workqueues.

* Not everyone runs in containers.  People running on bare metal or VMs can't be assumed to have a new enough version of cryptsetup, if that's a requirement.
* Any time I assume X will work for everyone or everyone will want X, someone pops up with a corner case
* Potential for other / future different arguments to `ceph-volume` for other purposes.

That said, if you have the code detect when it can disable workqueues and just do it, that'd be fine by me.  The sector size isn't on my radar.

baergj commented:
@anthonyeleven As it currently stands, the code does detect when cryptsetup has this option; if cryptsetup supports it and the kernel doesn't, cryptsetup will fall back to not disabling workqueues.

hydro-b commented:
> Given that Reef will likely be on CentOS 9, which comes with cryptsetup 2.4.3 as far as I can tell, my recommendation would be to drop the sector size logic from this PR and just focus on disabling the workqueues, unless you think that the cryptsetup 2.4+ sector size logic can be improved on? (In which case separating the PRs might make sense, since the benefit from the workqueue logic is more clear.)

I just checked Reef container, and it still is CentOS 8 Stream with cryptsetup 2.3.7. Not sure when the Ceph developers want to change to CentOS Stream 9, but the advice from the CentOS stream blog post to "migrate in time" makes sense: [CentOS Stream 8 EOL](https://blog.centos.org/2023/04/end-dates-are-coming-for-centos-stream-8-and-centos-linux-7/). EOL end of may 2024.

baergj commented:
@hydro-b Yeah, I talked to @markhpc about that last week - sounds like there were some bumps trying to get to 9 in the Reef timeframe. I think it's likely we'll see it for Squid (which is where this PR would now be introduced) given the May 2024 EOL.

markhpc commented:
> @hydro-b Yeah, I talked to @markhpc about that last week - sounds like there were some bumps trying to get to 9 in the Reef timeframe. I think it's likely we'll see it for Squid (which is where this PR would now be introduced) given the May 2024 EOL.

There may be a chance this gets backported to reef before that fwiw.

baergj commented:
To be clear, the underlying OS is not relevant to this PR in its current form; if there's still interest in the sector size optimization before we gain native cryptsetup support for it, I would argue that should still be a separate PR.

For this PR, I think we're either waiting for the ceph-volume maintainer to weigh in, or for someone else from Ceph with the authority to make the call on whether this is ready to merge. IMO the rotational detection still needs to be cleaned up but it's not ultimately my call.

hydro-b commented:
> To be clear, the underlying OS is not relevant to this PR in its current form; if there's still interest in the sector size optimization before we gain native cryptsetup support for it, I would argue that should still be a separate PR.

Agreed.

> For this PR, I think we're either waiting for the ceph-volume maintainer to weigh in, or for someone else from Ceph with the authority to make the call on whether this is ready to merge. IMO the rotational detection still needs to be cleaned up but it's not ultimately my call.

I also agree on this point. But before I make another attempt I would like to have some feedback from the maintainer, or someone else with authority to get some guidance and see how this can best be fixed.

markhpc commented:
@guits Any thoughts on this?  We're at a bit of a loss who should ultimately approve this.

anthonyeleven commented:
bump

frittentheke commented:
@cbodley could you maybe take a peek at this one?

cbodley commented:
> @cbodley could you maybe take a peek at this one?

sorry, i'm missing too much context to give any useful feedback here

markhpc commented:
@aclamk Did you have any luck pinging @guits?

guits commented on src/ceph-volume/ceph_volume/util/encryption.py:
```diff
@@ -23,11 +24,56 @@ def get_key_size_from_conf():
+    """
+    cryptsetup_help = execCmd("cryptsetup --help 2>/dev/null")
+
+    if '--perf-no_read_workqueue' not in cryptsetup_help or is_rotational(device):
```
```suggestion
    if '--perf-no_read_workqueue' not in cryptsetup_help or lsblk(device)['ROTA'] == '1':
```

As far as I know, they key 'ROTA' is always present but we might need to double-check that. If it can be missing, then maybe we would need a try/except KeyError sequence.

hydro-b replied:
@guits Thanks for reviewing this. I'll make time to address the issues where possible and see if we can make it so that it is fit for merging.

hydro-b commented on src/ceph-volume/ceph_volume/util/encryption.py:
```diff
@@ -23,11 +24,56 @@ def get_key_size_from_conf():
     return key_size
 
+
+def execCmd(cmdline):
+    cmd = subprocess.Popen(cmdline, shell=True, stdout=subprocess.PIPE)
+    stdout, _ = cmd.communicate()
+    return stdout.decode("utf-8")
```
Thanks for reviewing this! 

No specific reason, just one way I knew how to make it work. I'm definitely not a python expert. I'll try your suggestion.

hydro-b commented on src/ceph-volume/ceph_volume/util/disk.py:
```diff
@@ -195,6 +195,26 @@ def device_family(device):
+    for line in out:
+        device.append(_lsblk_parser(line))
+
+    for i in device:
+        if type(i) is dict and "ROTA" in i.keys():
+            if i['ROTA'] == '1':
+                rotational = True
+            else:
+                rotational = False
+
+    return rotational
```
It's been a while ago, but IIRC this will not work for LVM based devices, which is the default in ceph-volume (LVM). On the host the names can be:
/dev/mapper/osd.0-osd.0
/dev/dm-0
/dev/osd.0/osd.0

And those are treated in a different way than raw block devices. And passing one of these to disk.lsblk will not work for ROTA. But I will retest again with your suggestion and see if that is (still) true. If you can test yourself and see if that works for you or not ... might be the quickest way to see what goes wrong and where.

guits replied:
```python
>>> disk.lsblk('/dev/mapper/vg_test-lv_test')
{'NAME': 'vg_test-lv_test', 'KNAME': 'dm-0', 'PKNAME': '', 'MAJ:MIN': '252:0', 'FSTYPE': '', 'MOUNTPOINT': '', 'LABEL': '', 'UUID': '', 'RO': '0', 'RM': '0', 'MODEL': '', 'SIZE': '100G', 'STATE': 'running', 'OWNER': 'ceph', 'GROUP': 'ceph', 'MODE': 'brw-rw----', 'ALIGNMENT': '0', 'PHY-SEC': '512', 'LOG-SEC': '512', 'ROTA': '1', 'SCHED': '', 'TYPE': 'lvm', 'DISC-ALN': '0', 'DISC-GRAN': '512B', 'DISC-MAX': '2G', 'DISC-ZERO': '0', 'PARTLABEL': ''}
```

```python
>>> disk.lsblk('/dev/mapper/vg_test-lv_test')['ROTA']
'1'
>>> disk.lsblk('/dev/dm-0')['ROTA']
'1'
>>> disk.lsblk('/dev/vg_test/lv_test')['ROTA']
'1'
>>>

hydro-b replied:
In a couple of places in the code the key "ROTA" defaults to "1" if it cannot be determined for whatever reason. So to make sure it really checks the state of the device (instead of returning a default "1")  I would like to ask you if you can test with a SSD and / or  NVMe disk. Or to similate a non-rotational device in a virtual machine.

For the VM option with libvirt you would would need to add an extra attribute to the domain xml: "rotation_rate" (and poweroff / poweron the VM afterwards):

(virsh edit domain would let you do that in an editor). Example:

```
 </disk>
     ...
     ...
      </source>
      <target dev='vdb' bus='scsi' rotation_rate='1'/>
      <alias name='scsi0-0-2-0'/>
      <address type='drive' controller='0' bus='0' target='2' unit='0'/>
 </disk>
```

Note:  in libvirt a rotation_rate of 1 indicates _non_rotational_ as opposed to what is used in the linux kernel ("0"). See [libvirt domain format](https://libvirt.org/formatdomain.html).

Note: merely doing a "echo 0 > /sys/block/vdb/queue/rotational" won't work, as I observed this value gets reset when ceph-volume runs.

guits replied:
```console
# lsblk
NAME    MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
vda     253:0    0   11G  0 disk
`-vda1  253:1    0   10G  0 part /
vdb     253:16   0  200G  0 disk
vdc     253:32   0  200G  0 disk
vdd     253:48   0  200G  0 disk
vde     253:64   0  200G  0 disk
vdf     253:80   0  200G  0 disk
vdg     253:96   0  200G  0 disk
vdh     253:112  0  200G  0 disk
vdi     253:128  0  200G  0 disk
vdj     253:144  0  200G  0 disk
vdk     253:160  0  200G  0 disk
nvme0n1 259:0    0  200G  0 disk
nvme1n1 259:1    0  200G  0 disk
nvme2n1 259:2    0  200G  0 disk
```
```python
>>> from ceph_volume.util.device import Device
>>> Device('/dev/nvme0n1').rotational
False
>>> Device('/dev/vdk').rotational
True
>>>
```

hydro-b replied:
Sorry, I should have been a bit more specific and ask for LVM logical volume test. See below:

```
TEST root@dennisosd2:~$ lsblk
NAME          MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sda             8:0    0   10G  0 disk 
├─sda1          8:1    0   60M  0 part /boot/efi
└─sda2          8:2    0   10G  0 part 
  ├─LVM0-SWAP 253:1    0  976M  0 lvm  [SWAP]
  └─LVM0-ROOT 253:2    0    9G  0 lvm  /
sdb             8:16   0  100G  0 disk 
└─osd.4-osd.4 253:4    0  100G  0 lvm  
sdc             8:32   0  100G  0 disk 
└─osd.5-osd.5 253:3    0  100G  0 lvm  
vda           252:0    0  100G  0 disk 
└─osd.3-osd.3 253:0    0  100G  0 lvm
```

Those are all disks in a virtual machine. `/dev/sda` (OS), `/dev/sdb`, `/dev/vda` are rotational devices. `/dev/sdc` is a non-rotational device.  

### Rotational device tests:

```
TEST root@dennisosd2:~$ cat /sys/block/sdb/queue/rotational 
1
```
```
ls -lrt /dev/mapper/
total 0
crw------- 1 root root 10, 236 Nov  9 11:57 control
lrwxrwxrwx 1 root root       7 Nov  9 11:57 LVM0-ROOT -> ../dm-2
lrwxrwxrwx 1 root root       7 Nov  9 11:57 LVM0-SWAP -> ../dm-1
lrwxrwxrwx 1 root root       7 Nov  9 11:57 osd.4-osd.4 -> ../dm-4
lrwxrwxrwx 1 root root       7 Nov  9 11:57 osd.3-osd.3 -> ../dm-0
lrwxrwxrwx 1 root root       7 Nov  9 11:57 osd.5-osd.5 -> ../dm-3

cat /sys/block/dm-4/queue/rotational 
1

>>> from ceph_volume.util import disk
>>> disk.lsblk('/dev/sdb')['ROTA']
'1'
>>> disk.lsblk('/dev/osd.4/osd.4')['ROTA']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'ROTA'
>>> disk.lsblk('/dev/dm-4')['ROTA']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'ROTA'

>>> from ceph_volume.util.device import Device
>>> Device('/dev/sdb').rotational
Executable lvs not found on the host, will return lvs as-is
Executable pvs not found on the host, will return pvs as-is
Executable pvs not found on the host, will return pvs as-is
True
>>> Device('/dev/osd.4/osd.4').rotational
Executable lvs not found on the host, will return lvs as-is
True
>>> Device('/dev/dm-4').rotational
Executable lvs not found on the host, will return lvs as-is
True
```

### Non-rotational device tests

```
TEST root@dennisosd2:~$ cat /sys/block/sdc/queue/rotational 
0

TEST root@dennisosd2:~$ cat /sys/block/dm-3/queue/rotational 
0

>>> from ceph_volume.util import disk
>>> disk.lsblk('/dev/sdc')['ROTA']
'0'
>>> disk.lsblk('/dev/osd.5/osd.5')['ROTA']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'ROTA'
>>> disk.lsblk('/dev/dm-5')['ROTA']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'ROTA'


>>> from ceph_volume.util.device import Device
>>> Device('/dev/sdc').rotational
Executable lvs not found on the host, will return lvs as-is
Executable pvs not found on the host, will return pvs as-is
Executable pvs not found on the host, will return pvs as-is
False
```

Note that this is correct, it is indeed non-rotational

```
>>> Device('/dev/osd.5/osd.5').rotational
Executable lvs not found on the host, will return lvs as-is
True
```

Note that this is incorrect, it is the same device but now identified as a rotational device

```
>>> Device('/dev/dm-5').rotational
Executable lvs not found on the host, will return lvs as-is
 stderr: blkid: error: /dev/dm-5: No such file or directory
failed to determine if device /dev/dm-5 is BlueStore. device should not be used to avoid false negatives. err: [Errno 2] No such file or directory: '/dev/dm-5'
failed to determine if device /dev/dm-5 is BlueStore. device should not be used to avoid false negatives. err: [Errno 2] No such file or directory: '/dev/dm-5'
 stderr: Unknown device "/dev/dm-5": No such file or directory
True
```

Note that this is incorrect, it is the same device but now identified as a rotational device

Regarding logging lvs/pvs not found as an Executable ...

```
TEST root@dennisosd2:~$ which lvs
/usr/sbin/lvs
TEST root@dennisosd2:~$ /usr/sbin/lvs
  WARNING: PV /dev/sdc in VG osd.5 is using an old PV header, modify the VG to update.
  WARNING: PV /dev/sdb in VG osd.4 is using an old PV header, modify the VG to update.
  WARNING: PV /dev/vda in VG osd.3 is using an old PV header, modify the VG to update.
  LV    VG    Attr       LSize    Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  ROOT  LVM0  -wi-ao----    8.98g                                                    
  SWAP  LVM0  -wi-ao----  976.00m                                                    
  osd.3 osd.3 -wi-ao---- <100.00g                                                    
  osd.4 osd.4 -wi-ao---- <100.00g                                                    
  osd.5 osd.5 -wi-ao---- <100.00g

TEST root@dennisosd2:~$ which pvs
/usr/sbin/pvs
TEST root@dennisosd2:~$ /usr/sbin/pvs
  WARNING: PV /dev/sdc in VG osd.5 is using an old PV header, modify the VG to update.
  WARNING: PV /dev/sdb in VG osd.4 is using an old PV header, modify the VG to update.
  WARNING: PV /dev/vda in VG osd.3 is using an old PV header, modify the VG to update.
  PV         VG    Fmt  Attr PSize    PFree
  /dev/sda2  LVM0  lvm2 a--    <9.94g    0 
  /dev/sdb   osd.4 lvm2 a--  <100.00g    0 
  /dev/sdc   osd.5 lvm2 a--  <100.00g    0 
  /dev/vda   osd.3 lvm2 a--  <100.00g    0

TEST root@dennisosd2:~$ stat /usr/sbin/lvs
  File: /usr/sbin/lvs -> lvm
  Size: 3         	Blocks: 0          IO Block: 4096   symbolic link
Device: fd02h/64770d	Inode: 286530      Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-11-09 11:31:15.804000000 +0100
Modify: 2020-02-13 22:21:51.000000000 +0100
Change: 2022-04-01 16:58:53.800000000 +0200
 Birth: -

TEST root@dennisosd2:~$ ls -al /usr/sbin/lvs
lrwxrwxrwx 1 root root 3 Feb 13  2020 /usr/sbin/lvs -> lvm
TEST root@dennisosd2:~$ ls -al /usr/sbin/pvs
lrwxrwxrwx 1 root root 3 Feb 13  2020 /usr/sbin/pvs -> lvm
```

But let not get distracted here.

If the assumption holds that "disk.lsblk" and "Device" should always return a correct (in the sense of the "true" state of rotational) value of key "ROTA" , than according to my tests both are broken: disk.lsblk in a hard way, while "Device" is silently broken (and returning a default "True" in case something goes wrong). I believe in case of LVM we end up here in the code: [rotational](https://github.com/guits/ceph/blob/dmcrypt-optim/src/ceph-volume/ceph_volume/util/device.py#L388) and a default of "1" gets returned.

So, how did I test and what?
1)  git clone your ceph repo
2 checkout the dmcrypt-optim branch
3) On a Ubuntu 20.04 VM running ceph 16.2.11-1 the "/usr/lib/python3/dist-packages/ceph_volume" directory was replaced with the ceph_volume directory in the dmcrypt-optim branch.
4) performed the tests

Just to make sure I'm not making any errors here: Can you reproduce the tests I did with LVM logical volume on your test system?

guits replied:
@hydro-b ok, thanks for this clear and detailed report/testing 👍 
I'm gonna test the same in my environment to be sure we are not missing anything and try to shed the light on your findings.

guits replied:
@hydro-b 

```console
# lsblk
NAME                                                                                                  MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINT
vda                                                                                                   253:0    0   11G  0 disk
`-vda1                                                                                                253:1    0   10G  0 part  /rootfs
vdb                                                                                                   253:16   0  200G  0 disk               #  <---- rotational device
`-vg_ceph2-lv_ceph2                                                                                   252:5    0  100G  0 lvm
vdc                                                                                                   253:32   0  200G  0 disk
vdd                                                                                                   253:48   0  200G  0 disk
vde                                                                                                   253:64   0  200G  0 disk
vdf                                                                                                   253:80   0  200G  0 disk
vdg                                                                                                   253:96   0  200G  0 disk
vdh                                                                                                   253:112  0  200G  0 disk
vdi                                                                                                   253:128  0  200G  0 disk
vdj                                                                                                   253:144  0  200G  0 disk
vdk                                                                                                   253:160  0  200G  0 disk
nvme0n1                                                                                               259:0    0  200G  0 disk
`-ceph--c673f4bb--c777--4ef9--b2fa--4ceff9848ea0-osd--block--210bd35f--841a--4e68--bef5--a65e56d9c539 252:0    0  200G  0 lvm
  `-OdCzrL-7smV-SuBf-Y6EE-ZmTg-R1zt-2smArW                                                            252:2    0  200G  0 crypt
nvme1n1                                                                                               259:1    0  200G  0 disk
`-ceph--50ab7485--e020--4763--83b5--36a61b19c5a7-osd--db--bf239e44--08f0--4d26--9fe8--affd21d49bf9    252:1    0  200G  0 lvm
  `-4eMOki-yIdt-BJJQ-qr4J-6dfy-KFuE-47bgMy                                                            252:3    0  200G  0 crypt
nvme2n1                                                                                               259:2    0  200G  0 disk               #  <---- non-rotational device
`-vg_ceph-lv_ceph                                                                                     252:4    0  100G  0 lvm
```


```console
# ls -l /dev/mapper/
total 0
lrwxrwxrwx. 1 ceph ceph       7 Nov  9 09:42 4eMOki-yIdt-BJJQ-qr4J-6dfy-KFuE-47bgMy -> ../dm-3
lrwxrwxrwx. 1 ceph ceph       7 Nov  9 09:42 OdCzrL-7smV-SuBf-Y6EE-ZmTg-R1zt-2smArW -> ../dm-2
lrwxrwxrwx. 1 root root       7 Nov  9 09:41 ceph--50ab7485--e020--4763--83b5--36a61b19c5a7-osd--db--bf239e44--08f0--4d26--9fe8--affd21d49bf9 -> ../dm-1
lrwxrwxrwx. 1 root root       7 Nov  9 09:41 ceph--c673f4bb--c777--4ef9--b2fa--4ceff9848ea0-osd--block--210bd35f--841a--4e68--bef5--a65e56d9c539 -> ../dm-0
crw-------. 1 root root 10, 236 Nov  9 08:47 control
lrwxrwxrwx. 1 root root       7 Nov 13 12:04 vg_ceph-lv_ceph -> ../dm-4
lrwxrwxrwx. 1 root root       7 Nov 13 12:10 vg_ceph2-lv_ceph2 -> ../dm-5
```

## rotational device

```console
# cat /sys/block/vdb/queue/rotational
1
# cat /sys/block/dm-5/queue/rotational
1
```

```console
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/vdb')['ROTA']
'1'
>>> disk.lsblk('/dev/vg_ceph2/lv_ceph2')['ROTA']
'1'
>>> disk.lsblk('/dev/dm-5')['ROTA']
'1'
>>> Device('/dev/vdb').rotational
True
>>> Device('/dev/vg_ceph2/lv_ceph2').rotational
device is not part of ceph: </dev/vg_ceph2/lv_ceph2>
True
>>> Device('/dev/dm-5').rotational
True
```


## Non-rotational device

```console
# cat /sys/block/nvme2n1/queue/rotational
0
# cat /sys/block/dm-4/queue/rotational
0
```

```console
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/nvme2n1')['ROTA']
'0'
>>> disk.lsblk('/dev/vg_ceph/lv_ceph')['ROTA']
'0'
>>> disk.lsblk('/dev/dm-4')['ROTA']
'0'
>>> Device('/dev/nvme2n1').rotational
False
>>> Device('/dev/vg_ceph/lv_ceph').rotational
device is not part of ceph: </dev/vg_ceph/lv_ceph>
False
>>> Device('/dev/dm-4').rotational
False
```

guits replied:
@hydro-b 

You are running on `Ubuntu 20.04 VM` whereas I'm running on `CentOS Stream release 8`, hence all the log messages

```
Executable XXX not found on the host, will return lvs as-is` you are seeing.
```

As per LVM experts, LVM commands should never been run from within a container, so we had to introduce a behaviour in ceph-volume to make it run any LVM commands from the host using nsenter.

with Centos, LVM binaries are located in `/sbin`,
with Ubuntu, LVM binaries are located in `/usr/sbin`

I think the difference between our results is especially a consequence of this behaviour.
That being said, ceph-volume tries multiple paths on the host to find LVM binaries including `/usr/sbin` so it should work.
I'm going to deploy a new environment using Ubuntu and investigate that. I'll let you know ASAP.

guits replied:
@hydro-b re-tested with Ubuntu focal and everything worked as expected, same results as with my previous test with CentOS 🤷‍♂️

hydro-b replied:
Thanks for testing again. I'm at a loss here ...

FWIW, on ubuntu there is a symlink for `/sbin`:
```
ls -al /sbin
lrwxrwxrwx 1 root root 8 Apr  1  2022 /sbin -> usr/sbin
```
So whatever ceph-volume would execute it should work.

I have a cephadm based Reef cluster on Ubuntu 20.04 (docker as container runtime). I modified a 18.2.0 container (`sha256:a2d68ce89c8910cde20540d5d17874a94987118b939fae2e31d48177496ea9a2` with your ceph-volume branch and replaced the ceph_volume directory.  I commited changes, tagged, pushed to private repository, pointed cephadm to new (global) image, and redeployed osd containers. That works. Re-creating an OSD with ceph original container image with encryption also works. If I redeploy this container with the modified ceph-volume image (while the underlying dm-crypt device is already active) it works as well. However, if I restart the container host the OSD cannot be brought online:

ceph-osd.1.log:
```
2023-11-14T10:12:59.567+0000 7fc503c4f640  0 set uid:gid to 167:167 (ceph:ceph)
2023-11-14T10:12:59.567+0000 7fc503c4f640  0 ceph version 18.2.0 (5dd24139a1eada541a3bc16b6941c5dde975e26d) reef (stable), process ceph-osd, pid 7
2023-11-14T10:12:59.567+0000 7fc503c4f640  0 pidfile_write: ignore empty --pid-file
2023-11-14T10:12:59.567+0000 7fc503c4f640 -1 bluestore(/var/lib/ceph/osd/ceph-1/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-1/block: (2) No such file or directory
2023-11-14T10:12:59.567+0000 7fc503c4f640 -1 ESC[0;31m ** ERROR: unable to open OSD superblock on /var/lib/ceph/osd/ceph-1: (2) No such file or directoryESC[0m
2023-11-14T10:13:21.080+0000 7fd897df4640  0 set uid:gid to 167:167 (ceph:ceph
```

Ceph-volume logging:

```
[2023-11-14 11:33:11,764][ceph_volume.main][INFO  ] Running command: ceph-volume  activate --bad-option
[2023-11-14 11:33:13,485][ceph_volume.main][INFO  ] Running command: ceph-volume  activate --osd-id 1 --osd-uuid c2342c73-14c5-4c6a-a51a-94c1c273c2cf --no-systemd --no-tmpfs
[2023-11-14 11:33:13,487][ceph_volume.process][INFO  ] Running command: nsenter --mount=/rootfs/proc/1/ns/mnt --ipc=/rootfs/proc/1/ns/ipc --net=/rootfs/proc/1/ns/net --uts=/rootfs/proc/1/ns/
uts /usr/sbin/lvs --noheadings --readonly --separator=";" -a --units=b --nosuffix -S tags={ceph.osd_id=1,ceph.osd_fsid=c2342c73-14c5-4c6a-a51a-94c1c273c2cf} -o lv_tags,lv_path,lv_name,vg_nam
e,lv_uuid,lv_size
[2023-11-14 11:33:13,490][ceph_volume.process][INFO  ] stderr nsenter: failed to execute /usr/sbin/lvs: No such file or directory
[2023-11-14 11:33:14,743][ceph_volume.main][INFO  ] Running command: ceph-volume  lvm deactivate 1 c2342c73-14c5-4c6a-a51a-94c1c273c2cf
[2023-11-14 11:33:14,744][ceph_volume.process][INFO  ] Running command: nsenter --mount=/rootfs/proc/1/ns/mnt --ipc=/rootfs/proc/1/ns/ipc --net=/rootfs/proc/1/ns/net --uts=/rootfs/proc/1/ns/
uts /usr/sbin/lvs --noheadings --separator=";" -a -o lv_tags,lv_path,lv_name,vg_name,lv_uuid,lv_size -S lv_tags={ceph.osd_fsid=c2342c73-14c5-4c6a-a51a-94c1c273c2cf}
[2023-11-14 11:33:14,748][ceph_volume.process][INFO  ] stderr nsenter: failed to execute /usr/sbin/lvs: No such file or directory
[2023-11-14 11:33:14,748][ceph_volume.devices.lvm.deactivate][ERROR ] No data or block LV found for OSD1
```

So I decided to run some tests:

Running the ceph built 18.2.0 container image:

```
root@reef13:/root# docker run -it --privileged -v /dev:/dev -v /run/udev:/run/udev -v /sys:/sys -v /run/lvm:/run/lvm -v /run/lock/lvm:/run/lock/lvm -v /:/rootfs quay.io/ceph/ceph@sha256:a2d68ce89c8910cde20540d5d17874a94987118b939fae2e31d48177496ea9a2
[root@5534f72c86f3 /]# nsenter --mount=/rootfs/proc/1/ns/mnt --ipc=/rootfs/proc/1/ns/ipc --net=/rootfs/proc/1/ns/net --uts=/rootfs/proc/1/ns/uts /usr/sbin/lvs
nsenter: failed to execute /usr/sbin/lvs: No such file or directory

[root@5534f72c86f3 /]# lsblk 
NAME                                                                                                  MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sda                                                                                                     8:0    0   20G  0 disk 
|-sda1                                                                                                  8:1    0    1M  0 part 
`-sda2                                                                                                  8:2    0   20G  0 part 
  |-LVM0-swap                                                                                         253:1    0    1G  0 lvm  [SWAP]
  `-LVM0-root                                                                                         253:2    0   19G  0 lvm  /rootfs
sdb                                                                                                     8:16   0  100G  0 disk 
`-ceph--9fe2ad66--a719--408b--ad5d--31c810be869c-osd--block--c2342c73--14c5--4c6a--a51a--94c1c273c2cf 253:3    0  100G  0 lvm  
sdc                                                                                                     8:32   0  100G  0 disk 
`-ceph--55cd063e--f236--4dc5--b625--899075150b9a-osd--block--380f35c9--41cd--4989--a2af--db3da262d839 253:4    0  100G  0 lvm  
sdd                                                                                                     8:48   0  100G  0 disk 
`-ceph--7e4efd3d--b53e--4103--98da--836fd486c046-osd--block--ff1d83ea--f50d--4c9d--bf68--2970045168d2 253:0    0  100G  0 lvm  
sr0                                                                                                    11:0    1  364K  0 rom  
[root@5534f72c86f3 /]# python3
Python 3.6.8 (default, Oct 25 2023, 14:42:08) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-21)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/sdb')['ROTA']
'0'
>>> 
[root@5534f72c86f3 /]# ls /dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf 
/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf

[root@5534f72c86f3 /]# python3
Python 3.6.8 (default, Oct 25 2023, 14:42:08) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-21)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf')['ROTA']
'0'
>>> Device('/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf').rotational
True
```
Note that `disk.lsblk` is correct, but `Device().rotational` is not

Running the modified 18.2.0 container image with your branch:

```
root@reef13:/root# docker run -it --privileged -v /dev:/dev -v /run/udev:/run/udev -v /sys:/sys -v /run/lvm:/run/lvm -v /run/lock/lvm:/run/lock/lvm -v /:/rootfs reef.test.bit.nl:5000/guitsdmcrypt@sha256:aeb4b3ea16129b42688fa71164e83e4121c142d042dc047000b7c440ec64b07b

[root@81b292159e15 /]# nsenter --mount=/rootfs/proc/1/ns/mnt --ipc=/rootfs/proc/1/ns/ipc --net=/rootfs/proc/1/ns/net --uts=/rootfs/proc/1/ns/uts /usr/sbin/lvs
nsenter: failed to execute /usr/sbin/lvs: No such file or directory
[root@81b292159e15 /]# python3
Python 3.6.8 (default, Oct 25 2023, 14:42:08) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-21)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ceph_volume.util import disk
>>> from ceph_volume.util.device import Device
>>> disk.lsblk('/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf')['ROTA']
'0'
>>> Device('/dev/ceph-9fe2ad66-a719-408b-ad5d-31c810be869c/osd-block-c2342c73-14c5-4c6a-a51a-94c1c273c2cf').rotational
False
```
Note that both `disk.lsblk` and `Device().rotational` are correct.

So it seems a bug has been fixed between 18.2.0 and your branch (pretty much main), which is good news. However, I'm unable to get this running on my end. I might be overlooking something, not sure.

Would it be possible to build a container-image from your branch (ceph CI/CD)? I could try running that. Just to eliminate any user error I might be introducing in my tests.

Maybe @anthonyeleven and / or @baergj have time to test your branch and report their results. That might be interesting as well.

guits replied:
@hydro-b I'll share here any outcome(s) anyway but it feels to me this discussion is becoming too long/verbose. I'm available on slack (ceph-storage.slack.com) / or IRC (OFTC) to direct message if you want...

guits commented on src/ceph-volume/ceph_volume/util/encryption.py:
```diff
@@ -23,11 +24,56 @@ def get_key_size_from_conf():
+    linear           v1.4.0
+    error            v1.5.0
+
+    If dm-crypt does not have support for no_read_workqueue /
+    no_write_workqueue the options will be discarded and
+    logged by the linux kernel:
+
+    device-mapper: table: major:minor: crypt: Invalid feature arguments
+    device-mapper: ioctl: error adding target to table
+
+    The encrypted device will come online without the options active.
```
ok, this is not what I've understood from this comment, I thought rc would still be 0...
Regardless, unless we realize this is definitely the only option I would like to think a bit more about this before we go with this approach, but yeh, I agree this seems tricky to achieve this in a different way 😖

guits commented:
@hydro-b please, have a look at https://github.com/ceph/ceph/pull/54423/files

hydro-b commented:
> @hydro-b please, have a look at https://github.com/ceph/ceph/pull/54423/files (still a draft, 'work in progress')

@guits Thanks. I will take a look right away and test it in my test environment.

hydro-b commented:
> > > @hydro-b please, have a look at https://github.com/ceph/ceph/pull/54423/files (still a draft, 'work in progress')
> > 
> > 
> > @guits Thanks. I will take a look right away and test it in my test environment.
> 
> I tested the (re-)provisioning of an OSD and that works. With the caveat for that the work queue bypass code could not do it's job, so that part is not tested yet. But let's await your tests and see how we can go from there.

@guits has updated his patch and it's working for me now. At least I tested on a cephadm containerized environment, but I'm pretty sure it will work in non-containerized as well. I will test that tomorrow.

Short summary: at least, in the past (during my tests), there was a bug with determining rotational. That has been fixed in the meantime. There was one more issue with Guillaume's original patch and he has fixed that as well.

hydro-b commented:
@markhpc All tests have passed. @guits has fixed the last issue. I have re-tested it. All looks good to me. @guits let me know he thinks the PR is ready for merging into main. Do you have time to review and approve?

markhpc commented:
@hydro-b I can, but I don't think I can remove @guits block on the PR.  he'll need to approve it separately.

hydro-b commented:
I will close this PR in favor of PR 54423

markhpc commented:
@hydro-b ok, but please know that your efforts here are greatly appreciated and I want to make sure you get credit for your work on this PR.
